engine/src/flutter/lib/ui/painting/image_decoder_impeller.cc

```diff
@@ -75,7 +76,196 @@ bool IsWideGamut(const SkColorSpace* color_space) {
 }
 
-static std::optional<impeller::PixelFormat> ToPixelFormat(SkColorType type) {
+absl::StatusOr<ImageDecoderImpeller::ImageInfo> ToImageInfo(
+    const SkImageInfo& sk_image_format) {
+  const std::optional<impeller::PixelFormat> pixel_format =
+      ImageDecoderImpeller::ToPixelFormat(sk_image_format.colorType());
+  if (!pixel_format.has_value()) {
+    std::string decode_error(
+        std::format("Codec pixel format is not supported (SkColorType={})",
+                    static_cast<int>(sk_image_format.colorType())));
+    return absl::InvalidArgumentError(decode_error);
+  }
+  return {{
+      .size =
+          impeller::ISize(sk_image_format.width(), sk_image_format.height()),
+      .format = pixel_format.value(),
+  }};
+}
+
+SkColorType ChooseCompatibleColorType(SkColorType type) {
+  switch (type) {
+    case kRGBA_F32_SkColorType:
+      return kRGBA_F16_SkColorType;
+    default:
+      return kRGBA_8888_SkColorType;
+  }
+}
+
+SkAlphaType ChooseCompatibleAlphaType(SkAlphaType type) {
+  return type;
+}
+
+absl::StatusOr<SkColorType> ChooseCompatibleColorType(
+    ImageDecoder::PixelFormat format) {
+  switch (format) {
+    case ImageDecoder::PixelFormat::kR32G32B32A32Float:
+      return kRGBA_F32_SkColorType;
+    default:
+      return absl::InvalidArgumentError("unsupported target pixel format");
+  }
+}
+
+absl::StatusOr<SkImageInfo> CreateImageInfo(
+    const SkImageInfo& base_image_info,
+    const SkISize& decode_size,
+    bool supports_wide_gamut,
+    ImageDecoder::PixelFormat target_format) {
+  const bool is_wide_gamut =
+      supports_wide_gamut ? IsWideGamut(base_image_info.colorSpace()) : false;
+  SkAlphaType alpha_type =
+      ChooseCompatibleAlphaType(base_image_info.alphaType());
+  if (target_format != ImageDecoder::PixelFormat::kOptimal) {
+    absl::StatusOr<SkColorType> target_skia_color_type =
+        ChooseCompatibleColorType(target_format);
+    if (!target_skia_color_type.ok()) {
+      return target_skia_color_type.status();
+    }
+    return base_image_info.makeWH(decode_size.width(), decode_size.height())
+        .makeColorType(target_skia_color_type.value())
+        .makeAlphaType(alpha_type)
+        .makeColorSpace(SkColorSpace::MakeSRGB());
+  } else if (is_wide_gamut) {
+    SkColorType color_type = alpha_type == SkAlphaType::kOpaque_SkAlphaType
+                                 ? kBGR_101010x_XR_SkColorType
+                                 : kRGBA_F16_SkColorType;
+    return base_image_info.makeWH(decode_size.width(), decode_size.height())
+        .makeColorType(color_type)
+        .makeAlphaType(alpha_type)
+        .makeColorSpace(SkColorSpace::MakeSRGB());
+  } else {
+    return base_image_info.makeWH(decode_size.width(), decode_size.height())
+        .makeColorType(ChooseCompatibleColorType(base_image_info.colorType()))
+        .makeAlphaType(alpha_type)
+        .makeColorSpace(SkColorSpace::MakeSRGB());
+  }
+}
+
+struct DecodedBitmap {
```

Comment:
Nit: Since you have followed the absl::StaturOr pattern elsewhere, perhaps make this struct not contain a separate error field and just use absl::StatusOr (or optional).

Reply:
done